### PR TITLE
WIP: THNN

### PR DESCRIPTION
**Pull request by Roger-luo**
I'll start working on complex THNN. Merge this when all the layer is in complex. 

## Some Notes
The complex version of BP(backpropagation) would be:

For linear layers:

(This is in LaTeX)

```LaTeX
\delta^l = (\omega^l)^{\dag}\delta^{l+1}\\
\frac{\partial l}{\partial \conj{\omega}} = \delta^{l+1} (z^l)^{\dag}\\
\frac{\partial l}{\partial \conj{b}} = \delta^{l+1}
```

For activations (but not like PReLu, which has extra parameters)
```LaTeX
\delta^l = \delta^{l+1}\conj{f'(z^l)}
```

So mostly, just change transpose to hermitian and derivative of the activation function to its conjugate will work.

## References

- Akira Hirose. Complex-valued neural networks: theories and applications, volume 5. World
Scientific, 2003
- Tohru Nitta. Complex-valued neural networks. In Encyclopedia of Artificial Intelligence,
pages 361–366. IGI Global, 2009
- Akira Hirose. Proposal of fully complex-valued neural networks. In Neural Networks, 1992.
IJCNN., International Joint Conference on, volume 4, pages 152–157. IEEE, 1992.
- Nitzan Guberman. On complex valued convolutional neural networks. arXiv preprint arXiv:1602.09046, 2016.

## TODOs
- [x] Linear
- [x] Sigmoid
- [x] Tanh
- [ ] Abs
- [ ] ...

### CODE REVIEW

**Comment by PhilippPelz** [comment]
Nice

**Comment by elbamos** [comment]
@Roger-luo How can I help?
